packages/core/database/src/index.ts

```diff
@@ -96,12 +101,39 @@ class Database {
     this.logger = config.logger ?? console;
   }
 
+  #createConnection(connection: Knex.Config) {
+    return createConnection(connection, {
+      pool: { afterCreate: afterCreate(this) },
+    });
+  }
+
   async init({ models }: { models: Model[] }) {
+    // for function connections, create Knex connection asynchronously
+    if (typeof this.config.connection.connection === 'function') {
+      debug('Creating knex connection from function');
+      const conn = {
+        ...this.config.connection,
+        connection: await this.config.connection.connection(),
+      };
+      this.#connection = this.#createConnection(conn);
+    }
+
     this.metadata.loadModels(models);
     await validateDatabase(this);
+
+    debug('Database initialized');
+
     return this;
   }
 
+  get connection() {
```

Comment:
Just a check for this. In register the whole db part is not initialized. so what you would NOT be a breaking change. But if it is not loaded in by bootstrap It WILL be a breaking change.

Reply:
For connection object format, it would already be initialized because we do it in the constructor 100% of the time, synchronously, and Strapi creates the Database in its own constructor before any user code is run.

For connection function format, db is the very first thing we load in Strapi bootstrap before running any user bootstrap functions, so it should still be fine for the case you mention with seeding, although if you have a case you can test where it would fail, please post an example. But there are still potentially cases where it could fail, like if user code tries to access `connection[someProperty]` during register rather than waiting for bootstrap, which is why we have to consider the connection function format to be experimental.

Reply:
During register the whole strapi.db object does not exist.

Also just tested and on bootstrap it will always work so you are right this is not a breaking change

Reply:
🤔 It should, we do this in the Strapi constructor: https://github.com/strapi/strapi/blob/8ed749ecd9fe2851f95d5f48e9a4ec95327ff57a/packages/core/core/src/Strapi.ts#L272

Maybe it's not available for some reason though